bookkeeper-server/src/main/java/org/apache/bookkeeper/bookie/Bookie.java

```diff
@@ -1141,6 +1145,40 @@ private void addEntryInternal(LedgerDescriptor handle, ByteBuf entry,
         getJournal(ledgerId).logAddEntry(entry, ackBeforeSync, cb, ctx);
     }
 
+    /**
+     * Force write on the journal assigned to the given ledger.
+     * It works like a regular addEntry with ackBeforeSync=false but without really
+     * writing to disk.
+     */
+    private void forceLedgerInternal(LedgerDescriptor handle,
+                                     WriteCallback cb, Object ctx, byte[] masterKey)
+            throws IOException, BookieException {
+        long ledgerId = handle.getLedgerId();
+
+
+        if (masterKeyCache.get(ledgerId) == null) {
+            // Force the load into masterKey cache
+            byte[] oldValue = masterKeyCache.putIfAbsent(ledgerId, masterKey);
+            if (oldValue == null) {
+                // new handle, we should add the key to journal ensure we can rebuild
+                ByteBuffer bb = ByteBuffer.allocate(8 + 8 + 4 + masterKey.length);
+                bb.putLong(ledgerId);
+                bb.putLong(METAENTRY_ID_LEDGER_KEY);
+                bb.putInt(masterKey.length);
+                bb.put(masterKey);
+                bb.flip();
+
+                getJournal(ledgerId).logAddEntry(bb, false /* ackBeforeSync */, new NopWriteCallback(), null);
```

Comment:
Not sure how I missed, but 'sync' flag would have made standard here instead of ackBeforeSync.

If sync == true -> sync to disk before ack
if sync == false -> no need to sync before ack; 
May be we can change it when we get a chance.
@sijie ?

Reply:
If you fell strong I will change.

I think that currently the exceptional case is sync=false so having ackBeforeSync=true lets us identify more clearly this case.
From the client it is an opt in feature, and it makes sense that here we see the client is requiring a special treatment

Reply:
I prefer leaving ackBeforeSync in the scope of this patch, in order to keep it simple. We can do a follow up refactor